crates/primitives/src/lib.rs

```diff
@@ -200,33 +200,41 @@ pub trait Withdrawal: fmt::Debug {
 }
 
 /// Chunk related extension methods for Block
-/// FIXME: gate this behind scroll feature
+#[cfg(feature = "scroll")]
 pub trait BlockChunkExt {
-    /// Hash the header of the block
-    fn hash_da_header(&self, hasher: &mut impl tiny_keccak::Hasher);
     /// Hash the l1 messages of the block
-    fn hash_l1_msg(&self, hasher: &mut impl tiny_keccak::Hasher);
+    fn hash_l1_msg(&self, initial_queue_hash: &B256) -> B256;
 }
 
+#[cfg(feature = "scroll")]
 impl BlockChunkExt for RecoveredBlock<types::reth::Block> {
     #[inline]
-    fn hash_da_header(&self, hasher: &mut impl tiny_keccak::Hasher) {
-        hasher.update(&self.number.to_be_bytes());
-        hasher.update(&self.timestamp.to_be_bytes());
-        hasher.update(
-            &U256::from_limbs([self.base_fee_per_gas.unwrap_or_default(), 0, 0, 0])
-                .to_be_bytes::<{ U256::BYTES }>(),
-        );
-        hasher.update(&self.gas_limit.to_be_bytes());
-        hasher.update(&(self.body().transactions.len() as u16).to_be_bytes()); // FIXME: l1 tx could be skipped, the actual tx count needs to be calculated
-    }
-
-    #[inline]
-    fn hash_l1_msg(&self, hasher: &mut impl tiny_keccak::Hasher) {
+    fn hash_l1_msg(&self, initial_queue_hash: &B256) -> B256 {
         use reth_primitives_traits::SignedTransaction;
-        use types::consensus::Typed2718;
-        for tx in self.body().transactions.iter().filter(|tx| tx.ty() == 0x7e) {
-            hasher.update(tx.tx_hash().as_slice())
+        use tiny_keccak::Hasher;
+
+        let mut rolling_hash = *initial_queue_hash;
+        for tx in self
+            .body()
+            .transactions
+            .iter()
+            .filter(|tx| tx.is_l1_message())
+        {
+            let mut hasher = tiny_keccak::Keccak::v256();
+            hasher.update(rolling_hash.as_slice());
+            hasher.update(tx.tx_hash().as_slice());
+
+            hasher.finalize(rolling_hash.as_mut_slice());
+
+            // clear last 36 bits
+            // https://github.com/scroll-tech/da-codec/blob/d028c537b995acbdf9f3d6db84d138268f9691e6/encoding/da.go#L817-L830
+            rolling_hash.0[27] &= 0xF0;
```

Comment:
There was an implementation mismatch between contracts and da-codec here.

Finally what was agreed upon is that only 32 bits will be cleared. The fix PR for da-codec is here: https://github.com/scroll-tech/da-codec/pull/42